File: hw2/container.cpp
//
// Created by TOPAPEC on 08.10.2021.
//
//
// Created by TOPAPEC on 23.09.2021.
//

#include "container.h"
#include "shape.h"
#include <algorithm>

container::container() {
    length = 0;
    cont = new shape*[MAX_LENGTH];
}

container::~container() {
    clear();
}

void container::clear() {
    for (int i = 0; i < length; i++) {
        delete cont[i];
    }
    length = 0;
}

void container::in(FILE *fptr) {
    while (!feof(fptr)) {
        cont[length] = shape::generic_init_input(fptr);
        if (cont[length] != nullptr) {
            ++length;
        }
    }
}

void container::in_rnd(int size) {
    while (length < size) {
        cont[length] = shape::generic_init_rand();
        if (cont[length] != nullptr) {
            length++;
        }
    }
}

void container::output(FILE *fptr) {
    fprintf(fptr, "Container contains %d elements.\n", length);
    for (int i = 0; i < length; ++i) {
        fprintf(fptr, "%d : ", i);
        cont[i]->output(fptr);
    }
}

double container::perimeter_sum() {
    double sum = 0.0;
    for (int i = 0; i < length; ++i) {
        sum += cont[i]->perimeter();
    }
    return sum;
}

void container::shaker_sort() {
    int control = length - 1;
    int left = 0;
    int right = length - 1;
    do {
        for (int i = left; i < right; i++) {
            if (cont[i]->perimeter() > cont[i + 1]->perimeter()) {
                std::swap(cont[i], cont[i + 1]);
                control = i;
            }
        }
        right = control;
        for (int i = right; i > left; i--) {
            if (cont[i]->perimeter() < cont[i - 1]->perimeter()) {
                std::swap(cont[i], cont[i - 1]);
                control = i;
            }
        }
        left = control;
    } while (left < right);
}

File: hw3/Container.py
from Figure import Figure
from FigureFabric import FigureFabric


class Container:
    def __init__(self):
        self.cont = []

    def input(self, fptr) -> None:
        for line in fptr:
            self.cont.append(FigureFabric.generic_init_input(line))

    def input_random(self, num_figures) -> None:
        for i in range(num_figures):
            self.cont.append(FigureFabric.generic_init_random())
            print(self.cont[-1].print())

    def output(self, fptr) -> None:
        fptr.write(f"Container contains {len(self.cont)} elements.\n")
        for i, figure in enumerate(self.cont):
            fptr.write(f"{i}. ")
            figure.output(fptr)

    def perimeter_sum(self) -> float:
        sum = 0.0
        for figure in self.cont:
            sum += figure.perimeter()
        return sum

    def shaker_sort(self):
        cont = self.cont
        n = len(cont)
        swapped = True
        start = 0
        end = n - 1
        while swapped == True:
            swapped = False
            for i in range(start, end):
                if cont[i].perimeter() > cont[i + 1].perimeter():
                    cont[i], cont[i + 1] = cont[i + 1], cont[i]
                    swapped = True
            if swapped == False:
                break
            swapped = False
            end = end - 1
            for i in range(end - 1, start - 1, -1):
                if cont[i].perimeter() > cont[i + 1].perimeter():
                    cont[i], cont[i + 1] = cont[i + 1], cont[i]
                    swapped = True
            start = start + 1
